### PR TITLE
Add button component

### PR DESCRIPTION
**Pull request by nickcolley**
Adds a button component to be used across applications

Note: We do not consume GOV.UK Elements directly due to the naming conventions being leaky, in time this component will be a wrapper for the [GOV.UK Frontend](https://github.com/alphagov/govuk-frontend) project's button component.
 
 Global button styles seen at `app/assets/stylesheets/helpers/_buttons.scss`
 should be considered deprecated with this component replacing any instances.
 
 govuk-* namespace replaced with pub-c-* otherwise we will conflict with https://github.com/alphagov/govuk-frontend/blob/master/docs/components.md

government-frontend PR using this component: https://github.com/alphagov/government-frontend/pull/418

| Before   |
|---|
| Big |
|   big |
| Not so big |
| lessbig |
| Pretty small |
| kindasmall |
| Small |
| small |

| After |
|---|
| Big |
| big |
| Medium |
| medium |
| Small |
| small |
| iPhone 6 iOS9 |
| iphone6 ios9 |
| Samsung Galaxy S7 |
| samsunggalaxy7 |
| Safari 6 |
| safari6 |
| Chrome 57 |
| chrome57 |
| Firefox 56 |
| firefox56 |
| Edge 15 |
| edge15 |
| IE11 |
| ie11 |
| IE10 |
| ie10 |
| IE9 |
| ie9 |
| IE8 |
| ie8 |
| IE7 |
| ie7 |
| IE6 |
| ie6 |

Also extends component for use in the govspeak component

Component Guide

screen shot 2017-08-01 at 13 15 21

Hardcoded button example

www.gov.uk/access-to-work/apply

screen shot 2017-08-01 at 13 14 47

### CODE REVIEW

**Comment by fofr** [comment]
@nickcolley Sass linter errors – need to bump gem and update some other rules.

File: app/assets/stylesheets/govuk-component/_button.scss [hunk 1]
```diff
@@ -0,0 +1,41 @@
+.pub-c-button {
```
Comment (fofr) [inline app/assets/stylesheets/govuk-component/_button.scss]: Do we need to document what `pub-c` means? As it stands this is different to the other static components.

File: app/views/govuk_component/docs/button.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+body: |
+  Button text should be short and describe the action the button performs.
+
+  GOV.UK Elements has more information on how buttons should be used https://govuk-elements.herokuapp.com/buttons/
+
+  Note: We do not consume GOV.UK Elements directly due to the naming conventions being leaky,
+  in time this component will be a wrapper for the [GOV.UK Frontend](https://github.com/alphagov/govuk-frontend) project's button component.
+
+  Global button styles seen at `app/assets/stylesheets/helpers/_buttons.scss`
+  should be considered deprecated with this component replacing any instances.
```
Comment (fofr) [inline app/views/govuk_component/docs/button.yml]: Good docs :100:

File: app/assets/stylesheets/govuk-component/_button.scss (hunk 1 above)
Comment (fofr) [inline app/assets/stylesheets/govuk-component/_button.scss]: Nevermind, you did this 👍

**Comment by fofr** [comment]
* How should we add the `!important` for forcing button colour?
* How do we have a fixture showing specific link styles in container?

**Comment by fofr** [comment]
Is there a branch showing usage on government-frontend?

**Comment by nickcolley** [comment]
@fofr the color leakage is not an issue with this component since it sets color using `:link` which has a higher specificity than our leaky styles. 🙏 

(Would be great to have this as a test though)

**Comment by nickcolley** [comment]
@fofr I've updated this to address the concerns around the govspeak button that'll be removed from government-frontend. (Note: I've tested static with master government-frontend, and the cascading overrides work out OK so this wont mean we need to co-ordinate a deployment)

File: test/govuk_component/button_test.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  test "renders a button if href not set" do
+    render_component(text: "Start now")
+    assert_select "button.pub-c-button"
+    assert_select "a.pub-c-button", false
+  end
+
+  test "renders info text" do
+    render_component(text: "Start now", info_text: "Information text")
+    assert_select ".pub-c-button", text: "Start now"
+    assert_select ".pub-c-button__info-text", text: "Information text"
```
Comment (andysellick) [inline test/govuk_component/button_test.rb]: Should we also test that the info-text element is not rendered unless the info_text parameter is included?

File: app/assets/stylesheets/govuk-component/govspeak/_button.scss [hunk 4]
```diff
@@ -0,0 +1,12 @@
+
+/* im really sorry */
```
Comment (nickcolley) [inline app/assets/stylesheets/govuk-component/govspeak/_button.scss]: Accidentally committed that, I'm going to update with a better comment :P

File: app/views/govuk_component/docs/button.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  (This is a challenge to the reader :))
+accessibility_criteria: |
+  The button must:
+
+  - accept focus
+  - be focusable with a keyboard
+  - indicate when it has focus
+  - activate when focused and space is pressed
+  - activate when focused and enter is pressed
+  - have a role of button
```
Comment (nickcolley) [inline app/views/govuk_component/docs/button.yml]: I took this as to say, the role is defined by default if you use a button element, whereas the role for an anchor needs to be added by specifying the role attribute.

File: test/govuk_component/button_test.rb (hunk 3 above)
Comment (nickcolley) [inline test/govuk_component/button_test.rb]: Added a test to check defaults 👍

File: app/views/govuk_component/docs/button.yml (hunk 5 above)
Comment (andysellick) [inline app/views/govuk_component/docs/button.yml]: That makes sense, but might be worth clarifying.

Comment (nickcolley) [inline app/views/govuk_component/docs/button.yml]: I like @tvararu 's approach which is that the accessibility criteria does not refer to implementation details so I think this is fine for now.

**Pull request merged**